prov/rxm/src/rxm.h

```diff
@@ -371,8 +372,62 @@ void rxm_pkt_init(struct rxm_pkt *pkt);
 int rxm_ep_msg_mr_regv(struct rxm_ep *rxm_ep, const struct iovec *iov,
 		       size_t count, uint64_t access, struct fid_mr **mr);
 void rxm_ep_msg_mr_closev(struct fid_mr **mr, size_t count);
-struct rxm_buf *rxm_buf_get(struct rxm_buf_pool *pool);
-void rxm_buf_release(struct rxm_buf_pool *pool, struct rxm_buf *buf);
+
+#define RXM_BUF_GET(mr_local, pool)				\
+	(mr_local ? rxm_buf_get_ex(pool) : rxm_buf_get(pool))
+
+#define RXM_TX_BUF_GET(rxm_ep)						\
+	(struct rxm_tx_buf *)RXM_BUF_GET((rxm_ep)->msg_mr_local,	\
+					 &(rxm_ep)->tx_pool)
+
+#define RXM_RX_BUF_GET(rxm_ep)						\
+	(struct rxm_rx_buf *)RXM_BUF_GET((rxm_ep)->msg_mr_local,	\
+					 &(rxm_ep)->rx_pool)
+
+static inline
+struct rxm_buf *rxm_buf_get(struct rxm_buf_pool *pool)
+{
+	struct rxm_buf *buf;
+
+	fastlock_acquire(&pool->lock);
+	buf = util_buf_alloc(pool->pool);
+	if (OFI_UNLIKELY(!buf)) {
+		fastlock_release(&pool->lock);
+		return NULL;
+	}
+	memset(buf, 0, sizeof(*buf));
+	dlist_insert_tail(&buf->entry, &pool->buf_list);
+	fastlock_release(&pool->lock);
+	return buf;
+}
+
+static inline
+struct rxm_buf *rxm_buf_get_ex(struct rxm_buf_pool *pool)
+{
+	struct rxm_buf *buf;
+	void *mr;
+
+	fastlock_acquire(&pool->lock);
+	buf = util_buf_alloc_ex(pool->pool, (void **)&mr);
+	if (OFI_UNLIKELY(!buf)) {
+		fastlock_release(&pool->lock);
+		return NULL;
+	}
+	memset(buf, 0, sizeof(*buf));
+	dlist_insert_tail(&buf->entry, &pool->buf_list);
+	fastlock_release(&pool->lock);
+	buf->desc = fi_mr_desc((struct fid_mr *)mr);
+	return buf;
+}
```

Comment:
Does splitting the function really help? It seems the branch prediction would almost always be right.

Reply:
I don't see any significant impact on latency, but the number of instructions are reduced. Inline version of this function + w/o branches save almost 20 instructions it help us to decrease latency on 0.01 msec. The result is stable